src/schema/validate.rs

```diff
@@ -0,0 +1,731 @@
+                // Check required properties
+                if let Some(required_props) = required {
+                    for required_prop in required_props.iter() {
+                        if !object_value.contains_key(&Value::String(required_prop.clone())) {
+                            return Err(ValidationError::MissingRequiredProperty {
+                                property: required_prop.clone(),
+                                path: path.into(),
+                            });
+                        }
+                    }
+                }
+
+                // Handle discriminated subobjects (allOf with if/then)
+                // Validates against the appropriate variant schema based on discriminator field value
+                if let Some(discriminated_subobject) = discriminated_subobject {
+                    Self::validate_discriminated_subobject_with_base(
+                        object_value,
+                        discriminated_subobject,
+                        properties,
+                        additional_properties,
+                        path,
+                    )?;
+                } else {
+                    // Only validate regular object properties if no discriminated subobject exists
+                    // Validate each property
+                    for (prop_name, prop_value) in object_value.iter() {
+                        // First, ensure the property key is a string
+                        let prop_name_str = match prop_name {
+                            Value::String(string_key) => string_key,
+                            _ => {
+                                return Err(ValidationError::NonStringKey {
+                                    key_type: Self::value_type_name(prop_name),
+                                    path: path.into(),
+                                });
+                            }
+                        };
+
+                        // Create property path lazily using a closure
+                        let make_prop_path = || {
+                            if path.is_empty() {
+                                format!("[{prop_name_str}]")
+                            } else {
+                                format!("{path}.{prop_name_str}")
+                            }
+                        };
+
+                        if let Some(prop_schema) = properties.get(prop_name_str) {
+                            // Property is defined in schema, validate against it
+                            Self::validate_with_path(prop_value, prop_schema, &make_prop_path())
+                                .map_err(|e| ValidationError::PropertyValidationFailed {
+                                    property: prop_name_str.clone(),
+                                    path: path.into(),
+                                    error: Box::new(e),
+                                })?;
+                        } else if let Some(additional_schema) = additional_properties {
```

Comment:
>additional_properties [](http://example.com/codeflow?start=64&length=21)

say the user does create a schema where there's an object with additional properties. How will validation of the rule against that look?

For example, if I have a rule try to access myObject.propertyNotDefinedInTheSchema